{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import re\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import pandas as pd\n",
    "import time\n",
    "import uuid\n",
    "from jsonschema import validate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_save_read(url, file, reader=pd.read_csv, sep='\\t', **kwargs):\n",
    "  ''' Download file from {url}, save it to {file}, and subsequently read it with {reader} using pandas options on {**kwargs}.\n",
    "  '''\n",
    "  if not os.path.exists(file):\n",
    "    if os.path.dirname(file):\n",
    "      os.makedirs(os.path.dirname(file), exist_ok=True)\n",
    "    df = reader(url, sep=sep, index_col=None)\n",
    "    df.to_csv(file, sep=sep, index=False)\n",
    "  return pd.read_csv(file, sep=sep, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = \"Mammalia/Homo_sapiens\"\n",
    "url = 'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/{}.gene_info.gz'.format(organism)\n",
    "file = '{}.gene_info.tsv'.format(organism)\n",
    "\n",
    "ncbi_gene = fetch_save_read(url, file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20240411</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20240416</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20231010</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20240403</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20240403</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193451</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923215</td>\n",
       "      <td>trnD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Asp</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193452</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923216</td>\n",
       "      <td>trnP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193453</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923217</td>\n",
       "      <td>trnA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193454</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923218</td>\n",
       "      <td>COX1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>20230818</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193455</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923219</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193456 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #tax_id   GeneID    Symbol LocusTag                   Synonyms  \\\n",
       "0          9606        1      A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1          9606        2       A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2          9606        3     A2MP1        -                       A2MP   \n",
       "3          9606        9      NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4          9606       10      NAT2        -            AAC2|NAT-2|PNAT   \n",
       "...         ...      ...       ...      ...                        ...   \n",
       "193451   741158  8923215      trnD        -                          -   \n",
       "193452   741158  8923216      trnP        -                          -   \n",
       "193453   741158  8923217      trnA        -                          -   \n",
       "193454   741158  8923218      COX1        -                          -   \n",
       "193455   741158  8923219  16S rRNA        -                          -   \n",
       "\n",
       "                                                  dbXrefs chromosome  \\\n",
       "0       MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19   \n",
       "1       MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12   \n",
       "2       HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...         12   \n",
       "3       MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8   \n",
       "4       MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8   \n",
       "...                                                   ...        ...   \n",
       "193451                                                  -         MT   \n",
       "193452                                                  -         MT   \n",
       "193453                                                  -         MT   \n",
       "193454                                                  -         MT   \n",
       "193455                                                  -         MT   \n",
       "\n",
       "       map_location                         description    type_of_gene  \\\n",
       "0          19q13.43              alpha-1-B glycoprotein  protein-coding   \n",
       "1          12p13.31               alpha-2-macroglobulin  protein-coding   \n",
       "2          12p13.31  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3              8p22               N-acetyltransferase 1  protein-coding   \n",
       "4              8p22               N-acetyltransferase 2  protein-coding   \n",
       "...             ...                                 ...             ...   \n",
       "193451            -                            tRNA-Asp            tRNA   \n",
       "193452            -                            tRNA-Pro            tRNA   \n",
       "193453            -                            tRNA-Ala            tRNA   \n",
       "193454            -      cytochrome c oxidase subunit I  protein-coding   \n",
       "193455            -                              l-rRNA            rRNA   \n",
       "\n",
       "       Symbol_from_nomenclature_authority  \\\n",
       "0                                    A1BG   \n",
       "1                                     A2M   \n",
       "2                                   A2MP1   \n",
       "3                                    NAT1   \n",
       "4                                    NAT2   \n",
       "...                                   ...   \n",
       "193451                                  -   \n",
       "193452                                  -   \n",
       "193453                                  -   \n",
       "193454                                  -   \n",
       "193455                                  -   \n",
       "\n",
       "       Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                     alpha-1-B glycoprotein                   O   \n",
       "1                      alpha-2-macroglobulin                   O   \n",
       "2         alpha-2-macroglobulin pseudogene 1                   O   \n",
       "3                      N-acetyltransferase 1                   O   \n",
       "4                      N-acetyltransferase 2                   O   \n",
       "...                                      ...                 ...   \n",
       "193451                                     -                   -   \n",
       "193452                                     -                   -   \n",
       "193453                                     -                   -   \n",
       "193454                                     -                   -   \n",
       "193455                                     -                   -   \n",
       "\n",
       "                                       Other_designations  Modification_date  \\\n",
       "0       alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20240411   \n",
       "1       alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20240416   \n",
       "2                       pregnancy-zone protein pseudogene           20231010   \n",
       "3       arylamine N-acetyltransferase 1|N-acetyltransf...           20240403   \n",
       "4       arylamine N-acetyltransferase 2|N-acetyltransf...           20240403   \n",
       "...                                                   ...                ...   \n",
       "193451                                                  -           20200909   \n",
       "193452                                                  -           20200909   \n",
       "193453                                                  -           20200909   \n",
       "193454                     cytochrome c oxidase subunit I           20230818   \n",
       "193455                                                  -           20200909   \n",
       "\n",
       "       Feature_type  \n",
       "0                 -  \n",
       "1                 -  \n",
       "2                 -  \n",
       "3                 -  \n",
       "4                 -  \n",
       "...             ...  \n",
       "193451            -  \n",
       "193452            -  \n",
       "193453            -  \n",
       "193454            -  \n",
       "193455            -  \n",
       "\n",
       "[193456 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['protein-coding', 'pseudo', 'other', 'unknown', 'ncRNA', 'tRNA',\n",
       "       'rRNA', 'scRNA', 'snoRNA', 'snRNA', 'biological-region'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_gene.type_of_gene.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def maybe_split(record):\n",
    "    ''' NCBI Stores Nulls as '-' and lists '|' delimited\n",
    "    '''\n",
    "    if record in {'', '-'}:\n",
    "        return set()\n",
    "    return set(record.split('|'))\n",
    "\n",
    "def supplement_dbXref_prefix_omitted(ids):\n",
    "    ''' NCBI Stores external IDS with Foreign:ID while most datasets just use the ID\n",
    "    '''\n",
    "    for id in ids:\n",
    "        # add original id\n",
    "        yield id\n",
    "        # also add id *without* prefix\n",
    "        if ':' in id:\n",
    "            yield id.split(':', maxsplit=1)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_gene['All_synonyms'] = [\n",
    "    set.union(\n",
    "      maybe_split(gene_info['Symbol']),\n",
    "      maybe_split(gene_info['Symbol_from_nomenclature_authority']),\n",
    "      maybe_split(str(gene_info['GeneID'])),\n",
    "      maybe_split(gene_info['Synonyms']),\n",
    "      maybe_split(gene_info['Other_designations']),\n",
    "      maybe_split(gene_info['LocusTag']),\n",
    "      set(supplement_dbXref_prefix_omitted(maybe_split(gene_info['dbXrefs']))),\n",
    "    )\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "  ]\n",
    "\n",
    "synonyms, gene_id = zip(*{\n",
    "    (synonym, gene_info['GeneID'])\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "    for synonym in gene_info['All_synonyms']\n",
    "  })\n",
    "ncbi_lookup_syn = pd.Series(gene_id, index=synonyms)\n",
    "symbols, cap, gene_id = zip(*{\n",
    "    (gene_info['Symbol'], gene_info['Symbol'].upper(), gene_info['GeneID'])\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "  })\n",
    "ncbi_lookup_sym = pd.Series(gene_id, index=symbols)\n",
    "ncbi_lookup_sym_cap = pd.Series(gene_id, index=cap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "index_values = ncbi_lookup_syn.index.value_counts()\n",
    "ambiguous = index_values[index_values > 1].index\n",
    "ncbi_lookup_syn_disambiguated = ncbi_lookup_syn[(\n",
    "(ncbi_lookup_syn.index == ncbi_lookup_syn) | (~ncbi_lookup_syn.index.isin(ambiguous))\n",
    ")]\n",
    "sym_dict = ncbi_lookup_sym.to_dict()\n",
    "syn_dict_cap = ncbi_lookup_sym_cap.to_dict()\n",
    "syn_dict = ncbi_lookup_syn_disambiguated.to_dict()\n",
    "def gene_lookup(gene):\n",
    "    gene_id = sym_dict.get(gene)\n",
    "    if gene_id: return str(gene_id)\n",
    "    gene_id = syn_dict_cap.get(gene)\n",
    "    if gene_id: return str(gene_id)\n",
    "    return str(syn_dict.get(gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3105'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_lookup('HLA-A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name_mapper = {\n",
    "    \"KI67\": \"MKI67\",\n",
    "    \"C16ORF61\": \"CMC2\",\n",
    "    \"C20ORF46\": \"TMEM74B\",\n",
    "    \"C9ORF30\": \"MSANTD3\",\n",
    "    \"GUS\": \"GUSB\",\n",
    "    \"RPLPO\": \"RPLP0\",\n",
    "    \"TRFC\": \"TFRC\",\n",
    "    \"TSBP\": \"TSBP1\"\n",
    "}\n",
    "invalid_genes = set()\n",
    "def valid_gene(gene):\n",
    "    gene = gene_name_mapper.get(gene, gene)\n",
    "    if gene_lookup(gene) != 'None':\n",
    "        return True\n",
    "    invalid_genes.add(gene)\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = {}\n",
    "invalid_conditions = set()\n",
    "def get_condition_id(condition):\n",
    "\tif condition in conditions: return conditions\n",
    "\telse:\n",
    "\t\tres = requests.get(\"https://www.ebi.ac.uk/ols4/api/v2/entities?search=%s&ontologyId=doid\"%condition)\n",
    "\t\tresults = res.json()\n",
    "\t\tif results['numElements'] > 0:\n",
    "\t\t\tfor e in results[\"elements\"]:\n",
    "\t\t\t\t# print(e)\n",
    "\t\t\t\t# print(e[\"label\"])\n",
    "\t\t\t\t# print(e[\"synonym\"])\n",
    "\t\t\t\tif type(e[\"label\"]) == str and e[\"label\"].lower() == condition.lower() and 'DOID' in e[\"curie\"] or (\"synonym\" in e and condition.lower() in [d.lower() for d in e[\"synonym\"] if type(d) == str]):\n",
    "\t\t\t\t\tprint('here')\n",
    "\t\t\t\t\tconditions[condition] = {\n",
    "\t\t\t\t\t\t\"label\": e[\"label\"],\n",
    "\t\t\t\t\t\t\"onto_id\": e[\"curie\"],\n",
    "\t\t\t\t\t\t\"source\": e[\"ontologyPreferredPrefix\"]\n",
    "\t\t\t\t\t}\n",
    "\t\t\t\t\treturn conditions\n",
    "\t\tif condition not in conditions:\n",
    "\t\t\tinvalid_conditions.add(condition)\n",
    "\t\t\treturn None\n",
    "\t\t\t\n",
    "\n",
    "\t\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "patents = {\n",
    "\t\"WIPO\": \"https://patentscope.wipo.int/search/en/detail.jsf?docId=%s\",\n",
    "\t\"Google Patents\": \"https://patents.google.com/patent/%s/en\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker, diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker, prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "monitoring_biomarker\n",
      "prognostic_biomarker, monitoring_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker, diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker, prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker, prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker, prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker\n",
      "diagnostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n",
      "prognostic_biomarker\n"
     ]
    }
   ],
   "source": [
    "with open('data/PMC_biomarkers_051824.gmt', 'w') as fw:\n",
    "    with open('data/PMC_Gene_Set_Biomarkers-PMC_Biomarker_Sets.tsv') as o:\n",
    "        for line in o:\n",
    "            if 'evidence_source' in line:\n",
    "                continue\n",
    "            group, c, pubmed, evidence_source, biomarker_temp, best_biomarker_role, *genes = line.split(\"\\t\")\n",
    "            print(best_biomarker_role)\n",
    "            set_genes = set(genes)\n",
    "            for item in set_genes.copy():\n",
    "                if len(item) == 0:\n",
    "                    set_genes.remove(item)\n",
    "                if item == '\\n':\n",
    "                    set_genes.remove(item)\n",
    "            gs_str = '\\t'.join(set_genes)\n",
    "            fw.write(f\"{group}\\t\\t{gs_str}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_of_marker_roles(role_string):\n",
    "    role_array = role_string.split(',')\n",
    "    for idx, role in enumerate(role_array):\n",
    "        role_array[idx] = {'role': role.strip().replace(\"_biomarker\", \"\")}\n",
    "    return role_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Breast Carcinoma\n",
      "here\n",
      "Coronary artery disease\n",
      "here\n",
      "Developmental and epileptic encephalopathy 101\n",
      "here\n",
      "Rheumatoid Arthritis\n",
      "here\n",
      "Colorectal Cancer\n",
      "here\n",
      "Iron deficiency anemia\n",
      "here\n",
      "Iron deficiency anemia\n",
      "Iron deficiency anemia\n",
      "Iron deficiency anemia\n",
      "Anemia\n",
      "here\n",
      "Iron deficiency anemia\n",
      "Chronic kidney disease\n",
      "here\n",
      "Brain glioma\n",
      "here\n",
      "Brain glioma\n",
      "Depressive Disorder\n",
      "here\n",
      "Thyroid cancer\n",
      "here\n",
      "Chronic Hepatitis B\n",
      "here\n",
      "Liver Cirrhosis\n",
      "here\n",
      "Hepatocellular Carcinoma\n",
      "here\n",
      "Liver Cirrhosis\n",
      "Hepatocellular Carcinoma\n",
      "Hepatocellular Carcinoma\n",
      "Psoriatic Arthritis\n",
      "here\n",
      "Heart disease\n",
      "here\n",
      "Lung mixed small cell and squamous cell carcinoma\n",
      "here\n",
      "Osteosarcoma\n",
      "here\n",
      "Head and neck squamous cell carcinoma\n",
      "here\n",
      "Chronic Kidney Disease\n",
      "here\n",
      "Gastrointestinal system disease\n",
      "here\n",
      "Traumatic Brain Injury\n",
      "here\n",
      "Subacute delirium\n",
      "here\n",
      "Bladder Cancer\n",
      "here\n",
      "Postpartum Depression\n",
      "here\n",
      "Postpartum Depression\n",
      "Periodontitis\n",
      "here\n",
      "Alzheimer's Disease\n",
      "here\n",
      "Traumatic Brain Injury\n",
      "Malignant pleural mesothelioma\n",
      "here\n",
      "Pancreatic Cancer\n",
      "here\n",
      "Atrial Septal Defect\n",
      "here\n",
      "Cutaneous Leishmaniasis\n",
      "here\n",
      "Tongue Squamous Cell Carcinoma\n",
      "here\n",
      "Tongue Squamous Cell Carcinoma\n",
      "COVID-19\n",
      "here\n",
      "Hepatitis B\n",
      "here\n",
      "Drug psychosis\n",
      "here\n",
      "Ovarian Clear Cell Carcinoma\n",
      "here\n",
      "Colorectal Cancer\n",
      "Cholangiocarcinoma\n",
      "here\n",
      "Bladder Cancer\n",
      "Nephrolithiasis\n",
      "here\n",
      "Medullary Thyroid Carcinoma\n",
      "here\n",
      "Hepatocellular carcinoma\n",
      "here\n",
      "Bladder cancer\n",
      "here\n",
      "Cardiovascular system disease\n",
      "here\n",
      "Chronic obstructive pulmonary disease\n",
      "here\n",
      "Pulmonary Tuberculosis\n",
      "here\n",
      "Brain Glioma\n",
      "here\n",
      "Perlman syndrome\n",
      "here\n",
      "Rheumatoid arthritis\n",
      "here\n",
      "Chronic obstructive pulmonary disease\n",
      "IgA glomerulonephritis\n",
      "here\n",
      "Dermatomyositis\n",
      "here\n",
      "Lung non-small cell carcinoma\n",
      "here\n",
      "Prostate Cancer\n",
      "here\n",
      "Systemic lupus erythematosus\n",
      "here\n",
      "Parkinson's disease\n",
      "here\n",
      "Nasopharynx carcinoma\n",
      "here\n",
      "amyotrophic lateral sclerosis\n",
      "here\n",
      "Pulmonary emphysema\n",
      "here\n",
      "Stomach Cancer\n",
      "here\n",
      "Pancreatic ductal adenocarcinoma\n",
      "here\n",
      "Parkinson's disease\n",
      "Cervical Cancer\n",
      "here\n",
      "Ankylosing spondylitis\n",
      "here\n",
      "Multiple sclerosis\n",
      "here\n",
      "Psoriasis\n",
      "here\n",
      "Melanoma\n",
      "here\n",
      "Hepatocellular carcinoma\n",
      "Hypertension\n",
      "here\n",
      "Lupus Nephritis\n",
      "here\n",
      "Hepatocellular carcinoma\n",
      "Lung non-small cell carcinoma\n",
      "Cardiovascular system disease\n",
      "Kidney cortex necrosis\n",
      "here\n",
      "Breast Cancer\n",
      "here\n",
      "Atopic dermatitis\n",
      "here\n",
      "Abdominal obesity-metabolic syndrome\n",
      "here\n",
      "Congestive heart failure\n",
      "here\n",
      "Bladder cancer\n"
     ]
    }
   ],
   "source": [
    "gmt = []\n",
    "biomarkers = []\n",
    "with open('data/PMC_Gene_Set_Biomarkers-PMC_Biomarker_Sets.tsv') as o:\n",
    "    for line in o:\n",
    "        if 'evidence_source' in line:\n",
    "            continue\n",
    "        group, c, pubmed, evidence_source, biomarker_temp, best_biomarker_role, *genes = line.split(\"\\t\")\n",
    "        print(c)\n",
    "        conditions = get_condition_id(c)\n",
    "        if c in conditions:\n",
    "            biomarker = {\n",
    "                # \"biomarker_id\": \"GeneSet_%s\"%len(biomarkers),\n",
    "                \"best_biomarker_role\": list_of_marker_roles(best_biomarker_role),\n",
    "                \"condition\": {\n",
    "                    \"id\": conditions[c][\"onto_id\"],\n",
    "                    \"recommended_name\": {\n",
    "                        \"id\": conditions[c][\"onto_id\"],\n",
    "                        \"name\": conditions[c][\"label\"],\n",
    "                        \"resource\": \"Disease Ontology\",\n",
    "                        \"url\": \"http://disease-ontology.org/term/%s\"%conditions[c][\"onto_id\"],\n",
    "                    }\n",
    "                },\n",
    "            }\n",
    "            evidence_source = []\n",
    "            evidence_source.append({\n",
    "                \"id\": pubmed,\n",
    "                \"database\": \"PubMed\",\n",
    "                \"url\": \"https://pubmed.ncbi.nlm.nih.gov/%s\"%pubmed, \n",
    "                \"evidence_list\":[],\n",
    "                'tags':[]\n",
    "            })\n",
    "            # if patent_id:\n",
    "            #     evidence_source.append({\n",
    "            #         \"evidence_id\": patent_id,\n",
    "            #         \"database\": patent_source,\n",
    "            #         \"url\": patents[patent_source]%patent_id\n",
    "            #     })\n",
    "            biomarker[\"evidence_source\"] = evidence_source\n",
    "            biomarker_component = []\n",
    "            for gene in genes:\n",
    "                evidence_source = []\n",
    "                evidence_source.append({\n",
    "                \"id\": pubmed,\n",
    "                \"database\": \"PubMed\",\n",
    "                \"url\": \"https://pubmed.ncbi.nlm.nih.gov/%s\"%pubmed,\n",
    "                \"evidence_list\":[],\n",
    "                \"tags\": [{\"tag\":\"assessed_biomarker_entity\"}, {\"tag\":\"best_biomarker_role\"}]\n",
    "                })\n",
    "                gene = gene.strip().upper()\n",
    "                if valid_gene(gene):\n",
    "                    gene = gene_name_mapper.get(gene, gene)\n",
    "                    b = {\n",
    "                        \"biomarker\": biomarker_temp.replace(\"{gene}\", gene),\n",
    "                        \"assessed_biomarker_entity\": {\n",
    "                            \"recommended_name\": gene\n",
    "                        },\n",
    "                        \"assessed_biomarker_entity_id\": \"NCBI:\"+gene_lookup(gene),\n",
    "                        \"assessed_entity_type\": \"gene\",\n",
    "                        \"evidence_source\": evidence_source\n",
    "                    }\n",
    "                    biomarker_component.append(b)\n",
    "            \n",
    "            biomarker[\"biomarker_component\"] = biomarker_component\n",
    "            biomarkers.append(biomarker)\n",
    "# \"assessed_entity_type\": ncbi_gene[ncbi_gene.GeneID == int(gene_lookup(gene))][\"type_of_gene\"].to_list()[0] + \" gene\",\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(biomarkers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "schema = requests.get(\"https://raw.githubusercontent.com/biomarker-ontology/biomarker-partnership/main/schema/v0.3.3/biomarker_schema.json\").json()\n",
    "sample = requests.get(\"https://raw.githubusercontent.com/biomarker-ontology/biomarker-partnership/main/supplementary_files/sample_data_model_structures/v0.3/sample_biomarker.json\").json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "validate(instance=biomarkers, schema=schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('out/PMC_biomarker_sets.json', 'w') as o:\n",
    "     json.dump(biomarkers, o)\n",
    "with open('output/PMC_biomarker_sets.json', 'w') as o:\n",
    "     json.dump(biomarkers, o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "invalid_conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Breast Carcinoma\n",
      "Coronary artery disease\n",
      "Developmental and epileptic encephalopathy 101\n",
      "Rheumatoid Arthritis\n",
      "Colorectal Cancer\n",
      "Iron deficiency anemia\n",
      "Iron deficiency anemia\n",
      "Iron deficiency anemia\n",
      "Iron deficiency anemia\n",
      "Anemia\n",
      "Iron deficiency anemia\n",
      "Chronic kidney disease\n",
      "Brain glioma\n",
      "Brain glioma\n",
      "Depressive Disorder\n",
      "Thyroid cancer\n",
      "Chronic Hepatitis B\n",
      "Liver Cirrhosis\n",
      "Hepatocellular Carcinoma\n",
      "Liver Cirrhosis\n",
      "Hepatocellular Carcinoma\n",
      "Hepatocellular Carcinoma\n",
      "Psoriatic Arthritis\n",
      "Heart disease\n",
      "Lung mixed small cell and squamous cell carcinoma\n",
      "Osteosarcoma\n",
      "Head and neck squamous cell carcinoma\n",
      "Chronic Kidney Disease\n",
      "Gastrointestinal system disease\n",
      "Traumatic Brain Injury\n",
      "Subacute delirium\n",
      "Bladder Cancer\n",
      "Postpartum Depression\n",
      "Postpartum Depression\n",
      "Periodontitis\n",
      "Alzheimer's Disease\n",
      "Traumatic Brain Injury\n",
      "Malignant pleural mesothelioma\n",
      "Pancreatic Cancer\n",
      "Atrial Septal Defect\n",
      "Cutaneous Leishmaniasis\n",
      "Tongue Squamous Cell Carcinoma\n",
      "Tongue Squamous Cell Carcinoma\n",
      "COVID-19\n",
      "Hepatitis B\n",
      "Drug psychosis\n",
      "Ovarian Clear Cell Carcinoma\n",
      "Colorectal Cancer\n",
      "Cholangiocarcinoma\n",
      "Bladder Cancer\n",
      "Nephrolithiasis\n",
      "Medullary Thyroid Carcinoma\n",
      "Hepatocellular carcinoma\n",
      "Bladder cancer\n",
      "Cardiovascular system disease\n",
      "Chronic obstructive pulmonary disease\n",
      "Pulmonary Tuberculosis\n",
      "Brain Glioma\n",
      "Perlman syndrome\n",
      "Rheumatoid arthritis\n",
      "Chronic obstructive pulmonary disease\n",
      "IgA glomerulonephritis\n",
      "Dermatomyositis\n",
      "Lung non-small cell carcinoma\n",
      "Prostate Cancer\n",
      "Systemic lupus erythematosus\n",
      "Parkinson's disease\n",
      "Nasopharynx carcinoma\n",
      "amyotrophic lateral sclerosis\n",
      "Pulmonary emphysema\n",
      "Stomach Cancer\n",
      "Pancreatic ductal adenocarcinoma\n",
      "Parkinson's disease\n",
      "Cervical Cancer\n",
      "Ankylosing spondylitis\n",
      "Multiple sclerosis\n",
      "Psoriasis\n",
      "Melanoma\n",
      "Hepatocellular carcinoma\n",
      "Hypertension\n",
      "Lupus Nephritis\n",
      "Hepatocellular carcinoma\n",
      "Lung non-small cell carcinoma\n",
      "Cardiovascular system disease\n",
      "Kidney cortex necrosis\n",
      "Breast Cancer\n",
      "Atopic dermatitis\n",
      "Abdominal obesity-metabolic syndrome\n",
      "Congestive heart failure\n",
      "Bladder cancer\n"
     ]
    }
   ],
   "source": [
    "gmt = []\n",
    "biomarkers = []\n",
    "with open('data/PMC_Gene_Set_Biomarkers-PMC_Biomarker_Sets.tsv') as o:\n",
    "    for line in o:\n",
    "        if 'evidence_source' in line:\n",
    "            continue\n",
    "        group, c, pubmed, evidence_source, biomarker_temp, best_biomarker_role, *genes = line.split(\"\\t\")\n",
    "        print(c)\n",
    "        conditions = get_condition_id(c)\n",
    "        if c in conditions:\n",
    "            biomarker = {\n",
    "                \"biomarker_id\": \"GeneSet_%s\"%len(biomarkers),\n",
    "                \"best_biomarker_role\": list_of_marker_roles(best_biomarker_role),\n",
    "                \"condition\": {\n",
    "                    \"id\": conditions[c][\"onto_id\"],\n",
    "                    \"recommended_name\": {\n",
    "                        \"id\": conditions[c][\"onto_id\"],\n",
    "                        \"name\": conditions[c][\"label\"],\n",
    "                        \"resource\": \"Disease Ontology\",\n",
    "                        \"url\": \"http://disease-ontology.org/term/%s\"%conditions[c][\"onto_id\"],\n",
    "                    }\n",
    "                },\n",
    "            }\n",
    "            evidence_source = []\n",
    "            evidence_source.append({\n",
    "                \"id\": pubmed,\n",
    "                \"database\": \"PubMed\",\n",
    "                \"url\": \"https://pubmed.ncbi.nlm.nih.gov/%s\"%pubmed, \n",
    "                \"evidence_list\":[],\n",
    "                'tags':[]\n",
    "            })\n",
    "            # if patent_id:\n",
    "            #     evidence_source.append({\n",
    "            #         \"evidence_id\": patent_id,\n",
    "            #         \"database\": patent_source,\n",
    "            #         \"url\": patents[patent_source]%patent_id\n",
    "            #     })\n",
    "            biomarker[\"evidence_source\"] = evidence_source\n",
    "            biomarker_component = []\n",
    "            for gene in genes:\n",
    "                evidence_source = []\n",
    "                evidence_source.append({\n",
    "                \"id\": pubmed,\n",
    "                \"database\": \"PubMed\",\n",
    "                \"url\": \"https://pubmed.ncbi.nlm.nih.gov/%s\"%pubmed,\n",
    "                \"evidence_list\":[],\n",
    "                \"tags\": [{\"tag\":\"assessed_biomarker_entity\"}, {\"tag\":\"best_biomarker_role\"}]\n",
    "                })\n",
    "                gene = gene.strip().upper()\n",
    "                if valid_gene(gene):\n",
    "                    gene = gene_name_mapper.get(gene, gene)\n",
    "                    b = {\n",
    "                        \"biomarker\": biomarker_temp.replace(\"{gene}\", gene),\n",
    "                        \"assessed_biomarker_entity\": {\n",
    "                            \"recommended_name\": gene\n",
    "                        },\n",
    "                        \"assessed_biomarker_entity_id\": \"NCBI:\"+gene_lookup(gene),\n",
    "                        \"assessed_entity_type\": \"gene\",\n",
    "                        \"evidence_source\": evidence_source\n",
    "                    }\n",
    "                    biomarker_component.append(b)\n",
    "            \n",
    "            biomarker[\"biomarker_component\"] = biomarker_component\n",
    "            biomarkers.append(biomarker)\n",
    "# \"assessed_entity_type\": ncbi_gene[ncbi_gene.GeneID == int(gene_lookup(gene))][\"type_of_gene\"].to_list()[0] + \" gene\",\n",
    "\n",
    "\n",
    "with open('out/PMC_biomarker_sets_with_test_id.json', 'w') as o:\n",
    "     json.dump(biomarkers, o)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
